kong/clustering/control_plane.lua

```diff
@@ -0,0 +1,534 @@
+        return queue_semaphore:post(...)
+      end
+    }
+  end
+
+  -- if clients table is empty, we might have skipped some config
+  -- push event in `push_config_loop`, which means the cached config
+  -- might be stale, so we always export the latest config again in this case
+  if isempty(self.clients) or not self.deflated_reconfigure_payload then
+    _, err = handle_export_deflated_reconfigure_payload(self)
+  end
+
+  self.clients[wb] = queue
+
+  if self.deflated_reconfigure_payload then
+    local _
+    -- initial configuration compatibility for sync status variable
+    _, _, sync_status = self:check_configuration_compatibility(
+                              { dp_plugins_map = dp_plugins_map, })
+
+    table_insert(queue, RECONFIGURE_TYPE)
+    queue.post()
+
+  else
+    ngx_log(ngx_ERR, _log_prefix, "unable to send initial configuration to data plane: ", err, log_suffix)
+  end
+
+  -- how control plane connection management works:
+  -- two threads are spawned, when any of these threads exits,
+  -- it means a fatal error has occurred on the connection,
+  -- and the other thread is also killed
+  --
+  -- * read_thread: it is the only thread that receives websocket frames from the
+  --                data plane and records the current data plane status in the
+  --                database, and is also responsible for handling timeout detection
+  -- * write_thread: it is the only thread that sends websocket frames to the data plane
+  --                 by grabbing any messages currently in the send queue and
+  --                 send them to the data plane in a FIFO order. Notice that the
+  --                 PONG frames are also sent by this thread after they are
+  --                 queued by the read_thread
+
+  local read_thread = ngx.thread.spawn(function()
+    while not exiting() do
+      local data, typ, err = wb:recv_frame()
+
+      if exiting() then
+        return
+      end
+
+      if err then
+        if not is_timeout(err) then
+          return nil, err
+        end
+
+        local waited = ngx_time() - last_seen
+        if waited > PING_WAIT then
+          return nil, "did not receive ping frame from data plane within " ..
+                      PING_WAIT .. " seconds"
+        end
+
+      else
+        if typ == "close" then
+          ngx_log(ngx_DEBUG, _log_prefix, "received close frame from data plane", log_suffix)
+          return
+        end
+
+        if not data then
+          return nil, "did not receive ping frame from data plane"
+        end
+
+        -- dps only send pings
+        if typ ~= "ping" then
+          return nil, "invalid websocket frame received from data plane: " .. typ
+        end
+
+        ngx_log(ngx_DEBUG, _log_prefix, "received ping frame from data plane", log_suffix)
+
+        config_hash = data
+        last_seen = ngx_time()
+        update_sync_status()
+
+        -- queue PONG to avoid races
+        table_insert(queue, PONG_TYPE)
+        queue.post()
```

Comment:
```suggestion
        queue:post()
```

Reply:
No, `queue` is a simple wrapper for semaphore, it can not use `:`